library/core/src/cmp.rs

```diff
@@ -58,10 +58,21 @@ use self::Ordering::*;
 ///
 /// - **Transitive**: if `A: PartialEq<B>` and `B: PartialEq<C>` and `A:
 ///   PartialEq<C>`, then **`a == b` and `b == c` implies `a == c`**.
+///   This must also work for longer chains, such as when `A: PartialEq<B>`, `B: PartialEq<C>`,
+///   `C: PartialEq<D>`, and `A: PartialEq<D>` all exist.
 ///
 /// Note that the `B: PartialEq<A>` (symmetric) and `A: PartialEq<C>`
 /// (transitive) impls are not forced to exist, but these requirements apply
 /// whenever they do exist.
+/// The requirement for transitive chains has a subtle consequence: consider a crate `alpha`
+/// defining type `A1` and `A2`, and a crate `beta` defining a type `B`. If `beta` uses `alpha` as a
+/// dependency and provides trait implementations for both `A1: PartialEq<B>` and `B:
+/// PartialEq<A2>`, the transitivity rules don't apply, as an implementation `A1: PartialEq<A2>`
+/// does not exist; but at a later point `alpha` could decide to add such a `A1: PartialEq<A2>`
+/// implementation, in a way that would then make `beta`'s traits implementations violate
+/// transitivity. Effectively, `beta` is already making `A1` and `A2` comparable for equality even
+/// though `alpha` does not permit comparing them. Crate `beta` should generally refrain from doing
+/// that to avoid the potential for such conflicts.
```

Comment:
I used fairly weak wording here: "should generally refrain". I feel like we should take a more clear stance against such impls, but that seems like a bigger change and I wasn't sure what T-libs-api was most comfortable with. I also have no idea how common such problematic impls are in the ecosystem.

In prior discussion, someone suggested it would be reasonable to have a library type that implements both `PartialEq<i32>` and `PartialEq<i64>` (and their symmetric versions), which violates this possible. However if the standard library were to add such heterogeneous impls then obviously they would have like `left.into() == right`, so as long as the other crates respect that (which seems like a reasonable expectation), this case would actually still be fine. I wasn't sure how much I should go into that possibility here.